### PR TITLE
client/zcash: replace deprecated address methods with unified address parsing

### LINKED ISSUE #2134: zcash RPC deprecations and imminent removals
https://zcash.github.io/zcash/user/deprecation.html

Most notably getnewaddress

### PR DESCRIPTION
**Pull request by buck54321**
`getnewaddress` and `getrawchangeaddress` RPCs are deprecated. There is no way to get a transparent address directly. Instead, get a unified address, parse "receivers", and pick out the transparent address.

Also bumped min version to 5.4.2

Resolve #2134

### CODE REVIEW

File: client/asset/btc/btc.go [hunk 1]
```diff
@@ -312,6 +312,9 @@ type BTCCloneCFG struct {
 	// ZECStyleBalance is for clones that don't support getbalances or
 	// walletinfo, and don't take an account name argument.
 	ZECStyleBalance bool
+	// ZECUnifiedAddrs is for ZCash, which doesn't support getnewaddress, and
+	// has to extract the transparent address from a unified address.
+	ZECUnifiedAddrs bool
```
Comment (chappjc) [inline client/asset/btc/btc.go:317]: Looks to resolve https://github.com/decred/dcrdex/issues/2134

File: client/asset/btc/rpcclient.go [hunk 2]
```diff
@@ -543,6 +596,37 @@ func (wc *rpcClient) refundAddress() (btcutil.Address, error) {
...
 }
 
+type ZecAddressForAccountResult struct {
+	Account          uint32   `json:"account"`
+	DiversifierIndex uint32   `json:"diversifier_index"`
+	ReceiverTypes    []string `json:"receiver_types"`
+	Address          string   `json:"address"`
+}
+
+type ZecUnifiedReceivers struct {
```
Comment (chappjc) [inline client/asset/btc/rpcclient.go:606]: unexport and/or doc?

**Review by chappjc — COMMENTED** [review]
Is the 5.4.2 requirement enforced?  Also, is it strictly required by something in the product, or just the harness wallet dump files?

File: client/asset/btc/btc.go (hunk 1 above)
Comment (buck54321) [reply client/asset/btc/btc.go:317]: I edited this into the PR description, but you must have already been on the page.

File: client/asset/btc/rpcclient.go (hunk 2 above)
Comment (buck54321) [reply client/asset/btc/rpcclient.go:606]: Will unexport for now, but I suspect that I will need these exported for work with shielded pools.

**Comment by buck54321** [comment]
> Is the 5.4.2 requirement enforced? Also, is it strictly required by something in the product, or just the harness wallet dump files?

I did make it a requirement, though I can try to undo that for now if you want. From what I gather, there are some important fixes in the later 5.x releases.

**Comment by chappjc** [comment]
> > Is the 5.4.2 requirement enforced? Also, is it strictly required by something in the product, or just the harness wallet dump files?
> 
> I did make it a requirement, though I can try to undo that for now if you want. From what I gather, there are some important fixes in the later 5.x releases.

I suppose it's fine since it's not going to be an inconvenient upgrade for anyone, afaik.  Just wondering if there was a compatibility issue.

File: client/asset/btc/rpcclient.go [hunk 3]
```diff
@@ -168,6 +174,48 @@ func (wc *rpcClient) connect(ctx context.Context, _ *sync.WaitGroup) error {
...
+
+	var zerothFound bool
+	for _, acct := range accts {
+		if acct.Number == 0 {
+			zerothFound = true
+			break
+		}
+	}
+	if !zerothFound {
+		var res accountDetail
```
Comment (ukane-philemon) [inline client/asset/btc/rpcclient.go:208]: Could return early so it's more readable.

```
if zerothFound {
return nil // zec wallet has zeroth account
}

....
```

**Review by ukane-philemon — COMMENTED** [review]
utAck

**Review by JoeGruffins — APPROVED** [review]

**Comment by buck54321** [comment]
@chappjc if you'd prefer, I can incorporate the refactor from #2250 here.

**Comment by chappjc** [comment]
> @chappjc if you'd prefer, I can incorporate the refactor from #2250 here.

I'm by no means hung up on anything in these PRs, just working on release-scoped stuff right now.  However, yeah, ideally we'd not add something on one place in PR A and move/remove it in PR B (e.g. `methodZGetAddressForAccount`, `checkZecAccount`, etc.).  Basically I can just not bother reviewing `checkZecAccount` right?

File: dex/testing/dcrdex/harness.sh [hunk 4]
```diff
@@ -155,7 +155,7 @@ if [ $ZEC_ON -eq 0 ]; then
         {
             "base": "ZEC_simnet",
             "quote": "BTC_simnet",
-            "lotSize": 1000000,
+            "lotSize": 1000000000,
```
Comment (buck54321) [inline dex/testing/dcrdex/harness.sh:158]: ZEC lot size of 0.01 was really low, and I got a dust output redeeming BTC with a price of 0.01.

**Comment by buck54321** [comment]
> > @chappjc if you'd prefer, I can incorporate the refactor from #2250 here.
> 
> I'm by no means hung up on anything in these PRs, just working on release-scoped stuff right now, like BTC bonds. However, yeah, ideally we'd not add something on one place in PR A and move/remove it in PR B (e.g. `methodZGetAddressForAccount`, `checkZecAccount`, etc.). Basically I can just not bother reviewing `checkZecAccount` right?

Refactored

**Review by chappjc — APPROVED** [review]
LGTM, although there's one new code conflict.

Any luck with the JSON integer parsing issue you reported?

**Comment by buck54321** [comment]
> Any luck with the JSON integer parsing issue you reported?

They were aware. They're working on it. Looks on track. https://github.com/zcash/zcash/pull/6448

It doesn't affect us, yet. I am trying to add an account number argument to `z_listaccounts` (`DRAFT NOTE` in #2250) but I'm hitting the integer parsing issue, so that's on hold for the time being.

File: dex/testing/zec/harness.sh [hunk 5]
```diff
@@ -316,18 +316,20 @@ sleep 1
...
 tmux send-keys -t $SESSION:4 "./alpha generate 400${DONE}" C-m\; ${WAIT}
 
-#################################################################################
+################################################################################
 # Send gamma and delta some coin
 ################################################################################
 
 ALPHA_ADDR="tmEgW8c44RQQfft9FHXnqGp8XEcQQSRcUXD"
 BETA_ADDR="tmSog4freWuq1aC13yf1996fy4qXPmv3GTB"
-DELTA_ADDR=`./delta getnewaddress`
```
Comment (chappjc) [inline dex/testing/zec/harness.sh:329]: We require stuff like perl in different places, so I don't think requiring `jq` is too much to ask if we want to eventually use the newer `z_` RPCs

**Pull request merged**